src/interval.jl

```diff
@@ -33,85 +33,64 @@ A closed `Interval` can be constructed with the `..` infix constructor:
 
 ```julia
 julia> Dates.today() - Dates.Week(1) .. Dates.today()
-Interval{Date,Closed,Closed}(2018-01-24, 2018-01-31)
+Interval{Date}(2018-01-24, 2018-01-31, (Closed, Closed))
 ```
 
 See also: [`AnchoredInterval`](@ref)
 """
-struct Interval{T, L <: Bound, R <: Bound} <: AbstractInterval{T,L,R}
+struct Interval{T} <: AbstractInterval{T}
+    bounds::UInt8  # bounds comes first to allow incomplete construction for unbounded intervals
```

Comment:
Just curious if the intention is to have Intervals.jl use the IntervalSets.jl interface isn't this moving in the wrong direction? Or is the plan to update IntervalSets.jl in some way?

Reply:
Yes, the plan is to:

1. Subtype IntervalSets.AbstractInterval
2. Provide conversion and promotion rules for different interval types
3. Provide documentation on when to use each interval type and why
4. More clearly identify what should live in each package (ie: core API in IntervalSets.jl and collections of intervals in Intervals.jl)

In the end, most packages can just depend on a common API in IntervalSets.jl and just use duck-typing.

Yes, IntervalSets.jl may require some updates, but I'd like to narrow down what those should actually be. AFAIK, we didn't end up using a lot of the flexibility included with this package... maybe Beacon has used more of it? Can you clarify why you mean by "moving in the wrong direction"?

The argument for reverting back to storing bounds in a field rather than parameters is to differentiate the Intervals.Interval from the IntervalSets.Interval types.  If you're just working with individual intervals or all intervals have the same bounds then use IntervalSets.Interval. If you're working with large collections of Intervals where you might not have consistent bounds then use the Interval.jl type. It's the same reasoning as using AnchoredInterval if you know that all intervals have the same span. 

Again, the goal is to have the two main packages complement rather than compete with each other.

Reply:
> Can you clarify why you mean by "moving in the wrong direction"?

What I meant by this is that [IntervalSets.jl encodes the bound type information as type information](https://github.com/JuliaMath/IntervalSets.jl/blob/6027f24be5e7b1b900aefdd6bfa12ce3775feb26/src/interval.jl#L9) which was closer to what was done before this PR than with the bit mask. Now that I understand your plan more with supporting both types this makes more sense.

> Yes, IntervalSets.jl may require some updates, but I'd like to narrow down what those should actually be. AFAIK, we didn't end up using a lot of the flexibility included with this package.

I think one of those changes is to have IntervalSets.jl to move from `Interval{L,R,T}` to `Interval{T,L,R}`. If we're going to have multiple interval types the element type is going to come first for other implementations like you have here.

> The argument for reverting back to storing bounds in a field rather than parameters is to differentiate the Intervals.Interval from the IntervalSets.Interval types. If you're just working with individual intervals or all intervals have the same bounds then use IntervalSets.Interval. If you're working with large collections of Intervals where you might not have consistent bounds then use the Interval.jl type. It's the same reasoning as using AnchoredInterval if you know that all intervals have the same span.

I can get behind having both types of concrete interval exist for different use cases. Where I'm less sure is how these packages should be combined (like you mentioned)